packages/protobuf/src/transform/index.ts

```diff
@@ -0,0 +1,933 @@
+
+      _protoScalarsMap.set(program, scalarMap);
+    }
+    // Lazy initialize this map of known proto scalars.
+
+    return scalarMap;
+  }
+
+  function scalarToProto(t: Scalar): ProtoType {
+    const fullName = getTypeName(t);
+
+    const protoType = getProtoScalarsMap(program).get(t);
+
+    if (!protoType) {
+      reportDiagnostic(program, {
+        code: "unsupported-field-type",
+        messageId: "unknown-scalar",
+        format: {
+          name: fullName,
+        },
+        target: t,
+      });
+      return unreachable("unknown scalar");
+    }
+
+    return protoType;
+  }
+
+  function computeEffectiveModel(model: Model, anonymousModelName: string): Model | undefined {
+    if (effectiveModelCache.has(model)) return effectiveModelCache.get(model);
+
+    let effectiveModel = getEffectiveModelType(program, model);
+
+    if (effectiveModel.name === "") {
+      // Name the model automatically if it is anonymous
+      effectiveModel = program.checker.createAndFinishType({
+        ...model,
+        name: anonymousModelName,
+      });
+    }
+
+    if (!program.stateMap(state.externRef).has(effectiveModel)) {
+      visitModel(effectiveModel, model);
+    }
+
+    effectiveModelCache.set(model, effectiveModel);
+
+    return effectiveModel;
+  }
+  // #endregion
+
+  function checkForNamespaceCollisions(files: ProtoFile[]) {
+    const namespaces = new Set<string | undefined>();
+
+    for (const file of files) {
+      if (namespaces.has(file.package)) {
+        reportDiagnostic(program, {
+          code: "namespace-collision",
+          format: {
+            name: `"${file.package}"` ?? "<empty>",
+          },
+          target: file.source,
+        });
+      }
+
+      namespaces.add(file.package);
+    }
+  }
+
+  /**
+   * @param model - the Model to convert
+   * @returns a corresponding message declaration
+   */
+  function toMessage(model: Model): ProtoMessageDeclaration {
+    return {
+      kind: "message",
+      name: model.name,
+      reservations: program.stateMap(state.reserve).get(model),
+      declarations: [...model.properties.values()].map((f) => toMessageBodyDeclaration(f, model)),
+    };
+  }
+
+  /**
+   * @param property - the ModelProperty to convert
+   * @returns a corresponding declaration
+   */
+  function toMessageBodyDeclaration(
+    property: ModelProperty,
+    model: Model
+  ): ProtoMessageBodyDeclaration {
+    if (property.type.kind === "Union") {
+      // TODO: union must have at least one variant.
+      // TODO: union members must have field decorators.
+      // TODO: union must be anonymous
+      // TODO: union fields mustn't be arrays or maps.
+      // TODO: union variants must be string-named
+
+      // TODO: this isn't actually possible at all, because "union has named variants" and "union must be anonymous"
+      // are currently contradictory, so we can temporarily allow union decls with the understanding that they will be
+      // certainly inlined.
+
```

Comment:
Adding a comment here too for tracking what we dicussed:
create issues for those and link